rust/src/sans_io/read.rs

```diff
@@ -0,0 +1,1017 @@
+    ChunkRecord,
+    PaddingAfterChunk,
+    EndMagic,
+}
+use CurrentlyReading::*;
+
+struct ChunkState {
+    decompressor: Option<Box<dyn Decompressor>>,
+    uncompressed_data_hasher: Option<crc32fast::Hasher>,
+    compressed_remaining: u64,
+    uncompressed_len: u64,
+    padding_after_compressed_data: usize,
+    crc: u32,
+}
+
+#[derive(Default)]
+struct RWBuf {
+    data: Vec<u8>,
+    start: usize,
+    end: usize,
+    hasher: Option<crc32fast::Hasher>,
+}
+
+impl RWBuf {
+    fn new(instantiate_hasher: bool) -> Self {
+        Self {
+            hasher: if instantiate_hasher {
+                Some(crc32fast::Hasher::new())
+            } else {
+                None
+            },
+            ..Default::default()
+        }
+    }
+
+    // Marks some bytes of the un-read part as read.
+    fn mark_read(&mut self, read: usize) {
+        if let Some(hasher) = self.hasher.as_mut() {
+            hasher.update(&self.data[self.start..self.start + read]);
+        }
+        self.start += read;
+    }
+
+    // returns the length of the unread section.
+    fn len(&self) -> usize {
+        self.end - self.start
+    }
+
+    // returns an immutable view of the entire unread section.
+    fn unread(&self) -> &[u8] {
+        &self.data[self.start..self.end]
+    }
+
+    // returns a mutable view of the un-written part of the buffer, resizing as needed to ensure
+    // N bytes are available to write into.
+    fn tail_with_size(&mut self, n: usize) -> &mut [u8] {
+        let desired_end = self.end + n;
+        self.data.resize(desired_end, 0);
+        &mut self.data[self.end..]
+    }
+
+    // clears the RWBuf. Does not affect hasher state.
+    fn clear(&mut self) {
+        self.data.clear();
+        self.start = 0;
+        self.end = 0;
+    }
+}
+
+/// Options for initializing [`LinearReader`].
+#[derive(Debug, Default, Clone)]
+pub struct LinearReaderOptions {
+    /// If true, the reader will not expect the MCAP magic at the start of the stream.
+    pub skip_start_magic: bool,
+    /// If true, the reader will not expect the MCAP magic at the end of the stream.
+    pub skip_end_magic: bool,
+    /// If true, the reader will yield entire chunk records. Otherwise, the reader will decompress
+    /// and read into chunks, yielding the records inside.
+    pub emit_chunks: bool,
+    // Enables chunk CRC validation. Ignored if `prevalidate_chunk_crcs: true`.
+    pub validate_chunk_crcs: bool,
+    // Enables chunk CRC validation before yielding any messages from the chunk. Implies
+    // `validate_chunk_crcs: true`.
+    pub prevalidate_chunk_crcs: bool,
+    // Enables data section CRC validation.
+    pub validate_data_section_crc: bool,
+    // Enables summary section CRC validation.
+    pub validate_summary_section_crc: bool,
+}
+
+impl LinearReaderOptions {
+    pub fn with_skip_start_magic(self, skip_start_magic: bool) -> Self {
+        Self {
+            skip_start_magic,
+            ..self
+        }
+    }
```

Comment:
nit: you could save two lines if you wanted

```suggestion
    pub fn with_skip_start_magic(mut self, skip_start_magic: bool) -> Self {
        self.skip_start_magic = skip_start_magic;
        self
    }
```